315 view modes customized

jenlampton opened the pull request:
Fixes https://github.com/backdrop/backdrop-issues/issues/315

herbdool commented:
I'm only sorta aware of whole conversation around this but I like this one. Seems to be user friendly while still keeping customized view modes visible

docwilmot commented:
This UI looks great. I think this is the best so far. 
Still a few issues though:
- Saving a new view mode for User or Comment gives a `Page not found` error, with messgae `The requested page "/display" could not be found.` but the display saves.
- The enable button for Comment view modes does nothing.
- Cosmetic: beginning the description text for the view mode table with "When Comment items are viewed as part of a list, they are often displayed differently than when viewed alone on a full page." make it sound like this page is about lists, which may be confusing.
- Cosmetic: the mashing together of the two table header rows looks funny when there are no customized modes (perhaps some empty text or hide  it?):
![weirdnorow](https://cloud.githubusercontent.com/assets/1042283/21583482/ea9c545c-d050-11e6-87cd-05c8dcea3e0b.PNG)

jenlampton commented:
Thanks for the review @docwilmot ~!  I'll go test comments more thoroughly (and maybe add some more tests, too).  I think I know where that displays problem is coming from :) 

> This UI looks great. I think this is the best so far.

Woohoo!  I think we're narrowing it down to something we all agree on :)

> Saving a new view mode for User or Comment gives a Page not found error, with messgae The requested page "/display" could not be found. but the display saves.

Fixed.

> The enable button for Comment view modes does nothing.

Same with the reset link, I think it's the menu handling. edit: Fixed now.

> Cosmetic: beginning the description text for the view mode table with "When Comment items are viewed as part of a list, they are often displayed differently than when viewed alone on a full page." make it sound like this page is about lists, which may be confusing.

What if we turned it around and put full-page before list? `When Comment items are viewed alone on a full page they are often displayed differently than when viewed as part of a list. Each way an item can be displayed is called a View Mode.`

> Cosmetic: the mashing together of the two table header rows looks funny when there are no customized modes (perhaps some empty text or hide it?):

Ah yes. good catch. Removed the headers for sections that have fewer than 1 items.

quicksketch commented on core/modules/entity/entity.api.php:
```diff
@@ -423,3 +423,157 @@ function hook_entity_prepare_view($entities, $type) {
+ *     displayed in this view mode will reuse the 'default' display settings
+ *     by default (e.g. right after the module exposing the view mode is
+ *     enabled), but administrators can later use the Field UI to apply custom
+ *     display settings specific to the view mode.
```
Regarding the key "custom_settings". Could/should we rename this property with a name that indicates it's a boolean true/false value? For example "customized", "enabled", or "status"? The name "custom_settings" indicates to me that it would be an array of options.

quicksketch replied:
Actually, nevermind. Looks like "custom_settings" is an existing key in core already, so we can't change it. We're just exposing it in a hook now.

quicksketch replied:
Arg, there's some confusion around this variable key. Is it "custom settings" or "custom_settings"? The current code in core uses both (as does this PR). Which one is "right" (if either)?

jenlampton replied:
It's both! See underscores in [field_view_mode_settings()](https://api.backdropcms.org/api/backdrop/core%21modules%21field%21field.module/function/field_view_mode_settings/1) vs spaces in [entity_get_info()](https://api.backdropcms.org/api/backdrop/core%21modules%21entity%21entity.module/function/entity_get_info/1).  Maybe we create a follow-up issue? See https://github.com/backdrop/backdrop-issues/issues/2449

quicksketch commented on core/modules/entity/entity.api.php:
```diff
@@ -423,3 +423,157 @@ function hook_entity_prepare_view($entities, $type) {
+ *
+ * Note: This hook is invoked inside an implementation of
+ * hook_entity_info_alter() so care must be taken not to call anything that
+ * will result in an additional, and hence recurisve call to entity_get_info().
```
Grammar and spelling: "recurisve" typo. There should also be a comma after "call".

jenlampton replied:
fixed.

quicksketch commented on core/modules/entity/entity.api.php:
```diff
@@ -423,3 +423,157 @@ function hook_entity_prepare_view($entities, $type) {
+ *
+ * Note: This hook is invoked inside an implementation of
+ * hook_entity_info_alter() so care must be taken not to call anything that
+ * will result in an additional, and hence recurisve call to entity_get_info().
```
Grammar and spelling: "recurisve" typo. There should also be a comma after "call".

jenlampton replied:
fixed.

quicksketch commented on core/modules/entity/entity.api.php:
```diff
@@ -423,3 +423,157 @@ function hook_entity_prepare_view($entities, $type) {
+ * Change the view mode of an entity that is being displayed.
+ *
+ * @param string $view_mode
+ *   The view_mode that is to be used to display the entity.
```
"view_mode" should be "view mode" in the description.

quicksketch replied:
It should also be noted that this variable is passed by reference, and changing the value will change the view mode that is used by the calling function.

jenlampton replied:
fixed.

quicksketch commented on core/modules/entity/entity.api.php:
```diff
@@ -423,3 +423,157 @@ function hook_entity_prepare_view($entities, $type) {
+ * Act on a view mode before it is created or updated.
+ *
+ * @param string $view_mode
+ *   The view_mode that is to be used to display the entity.
```
"view_mode" => "view mode" in the description.

jenlampton replied:
fixed.

quicksketch commented on core/modules/entity/entity.api.php:
```diff
@@ -423,3 +423,157 @@ function hook_entity_prepare_view($entities, $type) {
+ * Respond to creation of a new view mode.
+ *
+ * @param string $view_mode
+ *   The view_mode that is to be used to display the entity.
```
"view_mode" => "view mode" in the description.

jenlampton replied:
Fixed

quicksketch commented on core/modules/entity/entity.api.php:
```diff
@@ -423,3 +423,157 @@ function hook_entity_prepare_view($entities, $type) {
+ * Respond to update of a view mode.
+ *
+ * @param string $view_mode
+ *   The view_mode that is to be used to display the entity.
```
"view_mode" => "view mode" in the description.

jenlampton replied:
fixed.

quicksketch commented on core/modules/entity/entity.api.php:
```diff
@@ -423,3 +423,157 @@ function hook_entity_prepare_view($entities, $type) {
+ * Respond to deletion of a view mode.
+ *
+ * @param string $view_mode
+ *   The view_mode that is to be used to display the entity.
```
"view_mode" => "view mode" in the description.

jenlampton replied:
fixed.

quicksketch commented on core/modules/entity/entity.api.php:
```diff
@@ -423,3 +423,157 @@ function hook_entity_prepare_view($entities, $type) {
+ * @param string $old_view_mode_name
+ *   The name of the view_mode before it is renamed.
+ * @param string $new_view_mode_name
+ *   The name of the view_mode after it is renamed.
```
"view_mode" => "view mode" in the descriptions.

jenlampton replied:
fixed.

quicksketch commented on core/modules/entity/entity.api.php:
```diff
@@ -423,3 +423,157 @@ function hook_entity_prepare_view($entities, $type) {
+ * @param string $new_view_mode_name
+ *   The name of the view_mode after it is renamed.
+ */
+function hook_entity_view_mode_rename($entity_type, $old_view_mode_name, $new_view_mode_name) {
```
I don't think we need to support renaming view modes. We don't support renaming taxonomy vocabularies, menus, fields, or just about anything that requires a machine name.

jenlampton replied:
works for me.

quicksketch commented on core/modules/entity/entity.module:
```diff
@@ -20,6 +20,192 @@ function entity_modules_disabled() {
+ * Load a custom entity view mode by entity type and machine name.
+ *
+ * @param (string) $entity_type
+ *   Entity type. For example, node, user or term.
```
Let's use a full sentence and accurate examples:

`The entity type of the view mode to be loaded, such as "node", "user", or "taxonomy_term".`

quicksketch replied:
No parenthesis around `(string)`

jenlampton replied:
really?  I've been doing it wrong!  (Fixed)

quicksketch commented on core/modules/entity/entity.module:
```diff
@@ -20,6 +20,192 @@ function entity_modules_disabled() {
+ *
+ * @param (string) $entity_type
+ *   Entity type. For example, node, user or term.
+ * @param  (string) $machine_name
```
No parenthesis around `(string)`

jenlampton replied:
fixed.

quicksketch commented on core/modules/entity/entity.module:
```diff
@@ -20,6 +20,192 @@ function entity_modules_disabled() {
+ * @param  (string) $machine_name
+ *   Machine name of view mode.
+ *
+ * @return  FALSE|(array)
```
No parenthesis around `(array)`. Extra space after `@return`.

jenlampton replied:
fixed.

quicksketch commented on core/modules/entity/entity.module:
```diff
@@ -20,6 +20,192 @@ function entity_modules_disabled() {
+ * @param (string) $entity_type
+ *   Entity type. For example, node, user or term.
+ * @param  (array) $view_mode
+ *   Fully loaded view mode, ready for saving.
```
Remove parentheses.

jenlampton replied:
fixed.

quicksketch commented on core/modules/entity/entity.module:
```diff
@@ -20,6 +20,192 @@ function entity_modules_disabled() {
+ */
+function entity_view_mode_save($entity_type, $view_mode) {
+  $existing_view_mode_name = !empty($view_mode['old_machine_name']) ? $view_mode['old_machine_name'] : $view_mode['machine_name'];
+  $view_mode_name = $view_mode['machine_name'];
```
As mentioned above, let's nix the renaming ability here.

jenlampton replied:
fixed.

quicksketch commented on core/modules/entity/entity.module:
```diff
@@ -20,6 +20,192 @@ function entity_modules_disabled() {
+ *
+ * @param (string) $entity_type
+ *   Entity type. For example, node, user or term.
+ * @param  (string) $machine_name
```
Same as other docblocks: remove parentheses, use full sentences, and there is an extra space after `@param` here.

jenlampton replied:
fixed,

quicksketch commented on core/modules/entity/entity.module:
```diff
@@ -20,6 +20,192 @@ function entity_modules_disabled() {
+ *
+ * @see entity_preprocess()
+ */
+function entity_entity_view($entity, $entity_type, $view_mode, $langcode) {
```
Let's type-hint the $entity parameter as `Entity $entity`.

jenlampton replied:
done

quicksketch commented on core/modules/entity/entity.module:
```diff
@@ -20,6 +20,192 @@ function entity_modules_disabled() {
+  $entity->content['#entity_view_mode'] = array(
+    'entity_type' => $entity_type,
+    'id' => $id,
+    'bundle' => $bundle,
```
Rather than use `entity_extract_ids()`, it's more efficient (and easier IMO) to use `$entity->id()` and `$entity->bundle()`.

jenlampton replied:
nice.

quicksketch commented on core/modules/entity/entity.module:
```diff
@@ -20,6 +20,192 @@ function entity_modules_disabled() {
+ * @see entity_view_mode_entity_view()
+ */
+function entity_preprocess(&$variables, $hook) {
+  // Check for the context provided in entity_view_mode_entity_view().
```
I couldn't find the function `entity_view_mode_entity_view()` to see why it is relevant. It's not in the PR or in core currently. Perhaps this is supposed to be `entity_entity_view()`?

I don't like that this is a generic `hook_preprocess()`. Is there no more specific preprocess function we can use? Otherwise this function will be called once for every theme() call, when it's really only relevant probably ~5% of the time.

quicksketch replied:
It might be possible to avoid this hook entirely if we just fix the calls to render entities directly, rather than altering in these suggestions. For example in `comment_view()` we have this:

```
  $build += array(
    '#theme' => 'comment__node_' . $node->type,
    '#comment' => $comment,
    '#node' => $node,
    '#view_mode' => $view_mode,
    '#language' => $langcode,
  );
```

If we changed `'#theme' => 'comment__node_' . $node->type,` to ` '#theme' => 'comment__node_' . $node->type . '__' . $view_mode`, we wouldn't need this preprocess hook.

Also... I'm not sure about offering `entity-type__id` and `entity-type__id__view-mode` suggestions globally. These are unlikely suggestions (or ones that should be discouraged). If needed, a theme can preprocess them back in.

jenlampton replied:
love it.

jenlampton replied:
added suggestions for comment, node, and term and user. Hope I didn't miss any entities.

quicksketch commented on core/modules/entity/entity.module:
```diff
@@ -20,6 +20,192 @@ function entity_modules_disabled() {
+/**
+ * Helper function: checks if the view mode exists.
+ */
+function entity_view_mode_exists($machine_name, $element, $form_state) {
```
This function doesn't seem to be called anywhere. Remove?

jenlampton replied:
It's called in `core/modules/field_ui/field_ui.admin.inc`.  It's a callback for the machine name field. Should we move it to field UI module?

quicksketch commented on core/modules/entity/entity.module:
```diff
@@ -522,3 +729,30 @@ function entity_autoload_info() {
+/**
+ * Implements hook_hook_info().
+ */
+function entity_hook_info() {
```
We haven't added other entity hooks in here. These should either be all or nothing IMO. Let's either add all entity hooks or remove this.

jenlampton replied:
removed.

quicksketch commented on core/modules/field/field.module:
```diff
@@ -1307,3 +1307,115 @@ function field_get_block_list() {
+ * Implements hook_entity_view_mode_rename().
+ */
+function field_entity_view_mode_rename($entity_type, $view_mode_old, $view_mode_new) {
+  $entity_info = entity_get_info($entity_type);
```
Remove renaming ability

jenlampton replied:
removed.

quicksketch commented on core/modules/field/field.module:
```diff
@@ -1307,3 +1307,115 @@ function field_get_block_list() {
+  // view mode for all bundles.
+  //if (empty($view_mode['custom settings'])) {
+  //  $view_mode['enabled_bundles'] = array_fill_keys(array_keys($entity_info['bundles']), 0);
+  //}
```
Should this be commented out?

jenlampton replied:
It should be removed, custom settings is not saved as part of the view mode in Backdrop, it's saved with field settings.

quicksketch commented on core/modules/field_ui/field_ui.admin.inc:
```diff
@@ -816,11 +816,77 @@ function field_ui_field_overview_form_submit($form, &$form_state) {
+    '#markup' => '<p>' . t('When @label items are viewed alone on a full page they are often displayed differently than when viewed as part of a list. Each way an item can be displayed is called a <strong>View Mode</strong>.', array('@label' => $bundle_label)) . '</p>',
+  );
+  $build['view_modes'] = array(
+    '#type' => 'group',
```
I don't think `#type = group` exists? We can probably remove this `#type` without any negative impact.

jenlampton replied:
removed.

quicksketch commented on core/modules/field_ui/field_ui.admin.inc:
```diff
@@ -1174,6 +1211,64 @@ function field_ui_display_overview_form($form, &$form_state, $entity_type, $bund
 }
 
+/**
+ * Form submission handler for field_ui_display_overview_form().
```
Docblock should say `field_ui_display_form()`, not `field_ui_display_overview_form()`

jenlampton replied:
fixed

quicksketch commented on core/modules/field_ui/field_ui.admin.inc:
```diff
@@ -2139,3 +2156,250 @@ function field_ui_next_destination($entity_type, $bundle) {
+}
+
+/**
+ * Submit hanlder for field_ui_view_mode_edit_form().
```
Typo: `hanlder`

`field_ui_view_mode_edit_form()` => `field_ui_view_mode_form()`

jenlampton replied:
fixed and fixed

klonos commented:
I like the updated version of this PR the most of all so far. I also spotted the issue with the customized/non-customized section headers when there is no view modes in them while playing around in the PR sandbox, but @docwilmot beat me to reporting it.

Perhaps this has to do with me playing around, but now the "Manage displays" page for post shows multiple "Customized for Post" headers. The "Manage displays" page for the Page content type is OK though.

quicksketch commented on core/modules/field_ui/field_ui.admin.inc:
```diff
@@ -2139,3 +2156,250 @@ function field_ui_next_destination($entity_type, $bundle) {
+      '%view-mode' => $view_mode['label']
+    )));
+
+    backdrop_goto($path);
```
Should use `$form_state['redirect'] = $path`. `backdrop_goto()` will incorrectly short-circuit any other submit handlers.

jenlampton replied:
ah, right. fixed.

quicksketch commented on core/modules/field_ui/field_ui.admin.inc:
```diff
@@ -2139,3 +2156,250 @@ function field_ui_next_destination($entity_type, $bundle) {
+    '%view-mode' => $form['#view_mode']['label'],
+  )));
+
+  backdrop_goto($form['#path']);
```
Use `$form_state['redirect']`

jenlampton replied:
done

quicksketch commented on core/modules/field_ui/field_ui.admin.inc:
```diff
@@ -2139,3 +2156,250 @@ function field_ui_next_destination($entity_type, $bundle) {
+  // Check to see if this view mode is in use on any other bundles.
+  $active = array();
+  $entity_info = entity_get_info($entity_type);
+  foreach($entity_info['bundles'] as $machine => $bundle) {
```
Space after `foreach`

jenlampton replied:
spaced

quicksketch commented on core/modules/field_ui/field_ui.admin.inc:
```diff
@@ -2139,3 +2156,250 @@ function field_ui_next_destination($entity_type, $bundle) {
+  }
+
+  if (!empty($active)) {
+    $message .= '<p>' . t('The <em>%view-mode</em> view mode has been customized in the following places:', array('%view-mode' => $view_mode['label'])) . '</p>';
```
There's no need for `<em>` tags if using a `%` replacement.

jenlampton replied:
nice catch

quicksketch commented on core/modules/field_ui/field_ui.admin.inc:
```diff
@@ -2139,3 +2156,250 @@ function field_ui_next_destination($entity_type, $bundle) {
+
+  return confirm_form(
+    $form,
+    t('Are you sure you want to Reset the %view-mode view mode?', array('%view-mode' => $view_mode['label'])),
```
`Reset` => `reset`

jenlampton replied:
fixed

quicksketch commented on core/modules/field_ui/field_ui.admin.inc:
```diff
@@ -2139,3 +2156,250 @@ function field_ui_next_destination($entity_type, $bundle) {
+  backdrop_set_message($message);
+
+  // Reload the current page.
+  backdrop_goto($path);
```
Use `$form_state['redirect']`.

jenlampton replied:
fixed

quicksketch commented on core/modules/field_ui/field_ui.module:
```diff
@@ -14,6 +14,17 @@ function field_ui_field_attach_rename_bundle($entity_type, $bundle_old, $bundle_
+function field_ui_permission() {
+  return array(
+    'administer view modes' => array(
+      'title' => t('Add, edit and delete custom view modes.'),
```
Oxford comma please :smile:

jenlampton replied:
love those <3

quicksketch commented on core/modules/field_ui/field_ui.module:
```diff
@@ -119,39 +131,77 @@ function field_ui_menu() {
+            'access arguments' => array('administer view modes'),
+            'file' => 'field_ui.admin.inc',
+          );
+          $items["$path/display/view-modes/%/reset"] = array(
```
Seems like all of these `%` placeholders should likely be `%entity_view_mode`, so that they get magic loaded into a view mode from `entity_view_mode_load()`. And it makes handling page not found easier if the view mode does not exist.

jenlampton replied:
Some view modes are provided by modules, and others are provided in config. The loader we have now only loads them from config, so that approach would fail for things like `full` and `teaser`.

quicksketch commented on core/modules/field_ui/tests/field_ui.test:
```diff
@@ -832,3 +831,291 @@ class FieldUIAlterTestCase extends BackdropWebTestCase {
+/**
+ * Tests view mode altering theme suggestions.
+ */
+class FieldUIViewModeSuggestionsTest extends FieldUIViewModeTestHelper {
```
Even if we are able to remove `entity_preprocess()`, we should keep the tests here that are still covered without it.

jenlampton replied:
👍

jenlampton replied:
oi, those tests completely failed without some reworking to get all the theme hook suggestions in place.

jenlampton replied:
these are now normal theme hook suggestions. Removing tests since we are not altering the way they work :)

jenlampton commented:
> Perhaps this has to do with me playing around, but now the "Manage displays" page for post shows multiple "Customized for Post" headers. The "Manage displays" page for the Page content type is OK though.

Nope, that was me. FIxed now! :)

quicksketch commented:
Testbot is having some trouble with random fails, but I've seen both 5.3 and 7.0 pass, so I'm merging this. We still have an issue for tracking this particular random failure at https://github.com/backdrop/backdrop-issues/issues/2202.

quicksketch commented:
Merged into 1.x for 1.6.0. Woot!
